feat(migration): Diff local schema with dataset schema

lynnagara opened the pull request:
Parse DESCRIBE TABLE output into a schema which can be diffed with the
table writer schema.

Also fixes incorrectly logged errors in local migrations. These were being
incorrectly compared when there were default expressions or codecs in
the column definition.

lynnagara committed:
fix(migration): Attempt to fix migration errors

Fix incorrectly logged errors in migrations. These were being
incorrectly compared when there were default expressions or codecs in
the column definition.

fpacifici commented on snuba/datasets/schemas/__init__.py:
```diff
@@ -107,10 +117,32 @@ def get_column_differences(self, expected_columns: Mapping[str, str]) -> List[st
+
+        if codec_expr:
+            column_type_expr = "{} CODEC ({})".format(column_type_expr, codec_expr)
+
```
Our migration system is so broken that it is hard to give a constructive feedback on how this should look like without 3 months of refactoring, still I think we can make a step in the right direction by start using ColumnType classes and subclasses instead of comparing strings.

The ColumnType abstraction is far from being a good one to manage migrations, but it is already present in the schema of the column and it would give you some structure at least.
So I suggest that, instead of fetching the ColumnType for the column from the schema, formatting it into string and comapring with a tweaked string coming from the `DESCRIBE` call, you parse the string coming from DESCRIBE into a ColumnType (that can be nested) and then compare it with the one present in the schema. 

Since this is not time critical now, I think it would be a good investment.

lynnagara replied:
Tackled this first since I think having a reliable way to compare current database state with the schema of a dataset will be generally useful.

I think using our standard ColumnTypes (and doing away with MigrationSchemaColumn completely) will be good. Probably don't need a separate column type representation just for migration.

It seems like it will be quite a bit harder than comparing the strings though, since:
- We need to store another set of mappings to go from string column descriptions back into ColumnType structures. Going the other way is easier because ColumnType already implements the `for_schema()` method for this.
- There might be more than one way to represent columns in our schema, so it's ambiguous what the schema should actually be
- One case I can think of that is like this is our Nested column type where we use the `.` notation as a convention but it the same output can possibly be represented in multiple ways in the dataset

fpacifici replied:
It seems both your first and second point apply equally to string comparison int hat you have to tweak the representations exactly in this method. I think we should think this problem through before adding a hack to avoid the warnings. 
Is there anything that would be blocked on this PR?

lynnagara replied:
No blockers. What do you mean that they apply equally though? My point was the opposite; I don't think those points apply to string comparison at all.
- Mapping from column type to string already exists (this is `for_schema()`), mapping from string to ColumnType needs to be added and is only needed if we are comparing the actual ColumnType instances
- There is a 1:many relationship between ColumnType: string output. Columns defined in schema always yield the same string in Clickhouse. The opposite isn't true, given a `DESCRIBE TABLE` output there are many ways it can be represented in the schema. (Nested vs Array is the one example of this)

lynnagara replied:
One other thing that is not very clear is how column type order is supposed to affect the query. For example is `WithCodec(WithDefault(x))` supposed to yield the same result as `WithDefault(WithCodec(x))`? Is one of these supposed to be invalid? Or are they meant to be smart enough to figure out what the right statement should be regardless of order applied?

lynnagara replied:
Last commit in this branch is the version with ColumnType and removed the MigrationSchemaColumn, there's a bit more code involved here

fpacifici replied:
```
For example is WithCodec(WithDefault(x)) supposed to yield the same statement as WithDefault(WithCodec(x))
```

That's a good question. I am not sure, could you please verify if clickhouse expects one form only ?

lynnagara replied:
I tried both and only the first variant - `WithCodecs(WithDefault(x, def), codecs)`ends up getting formatted correctly by Clickhouse. But I guess my point wasn't about what happens today (which I'd expect to be not 100% correct) but what we should be doing if a user provides the reverse `WithDefault(withCodecs(x))`. Should we throw an error immediately? Or just accept it and transform it into the right thing? What if there are other examples out there that are both accepted by Clickhouse? Should we always enforce one "correct" order to apply the column types?

fpacifici replied:
```
but what we should be doing if a user provides the reverse WithDefault(withCodecs(x)). Should we throw an error immediately? 
```
That's my question: does clickhouse accept both ? And if it accepts both does it consider them equivalent in a DESCRIBE TABLE statement ?

lynnagara replied:
Clickhouse only accepts one. I suppose they are equivalent in that they are completely separate columns in the describe table output

lynnagara committed:
Merge remote-tracking branch 'origin/master' into fix-migrate-errors

lynnagara committed:
use column type to compare schema, remove MigrationSchemaColumn

lynnagara commented on snuba/datasets/schemas/tables.py:
```diff
@@ -41,12 +41,6 @@ def get_prewhere_candidates(self) -> Sequence[str]:
         return self.__prewhere_candidates
 
 
-class MigrationSchemaColumn(NamedTuple):
```
It would have caused circular dependencies since we needed to use it in snuba/datasets/schemas/__init__.py (in the previous version that did not remove it completely)

fpacifici commented on snuba/datasets/transactions.py:
```diff
@@ -137,10 +135,14 @@ def transactions_migrations(
+            if isinstance(col, WithDefault):
+                col.inner_type = LowCardinality(col.inner_type)
+            else:
+                col = LowCardinality(col)
```
Why not just apply the alter without doing any check ?
It should be a noop if the column is already in the right format. But I would verify on Clickhouse if it complains bout this.

lynnagara replied:
Do you mean making this potentially `LowCardinality(LowCardinality(String()))`? That won't work.

fpacifici replied:
No, nevermind, I thought about running something like 
```
ALTER TABLE table MODIFY COLUMN {col_name} {col_type}
```
where col_type is the formatted expression from the Snuba schema (the source of truth), not from the current schema. But I realized you do not have access to the snuba schema here.

fpacifici commented on snuba/migrate.py:
```diff
@@ -14,11 +34,84 @@ def _run_schema(conn, schema):
+        match = STRIP_CAST_RE.match(default_expr)
+        if match:
+            default_expr = match.groups()[0]
+        return default_expr
```
default_expr is not declared if the regex did not match anything. Did mypy complain ?

lynnagara replied:
It just returns the untransformed default_expr if the regex doesn't match

fpacifici commented on snuba/migrate.py:
```diff
@@ -14,11 +34,84 @@ def _run_schema(conn, schema):
         return
     clickhouse_table = schema.get_local_table_name()
 
+    STRIP_CAST_RE = re.compile(r"^CAST\((.*), (.*)\)$", re.IGNORECASE)
```
we need unit tests for this logic

fpacifici commented on snuba/migrate.py:
```diff
@@ -14,11 +34,84 @@ def _run_schema(conn, schema):
+    ENUM_VALUES_RE = re.compile(r"'(.+)' = (\d+)$")
+    MODIFIER_COL_RE = re.compile(r"^(Array|LowCardinality|Nullable)\((.+)\)$")
+
+    def strip_cast(default_expr: str) -> str:
```
Have you consider writing a grammar for this? It seems it would be easier to extend and manage than the ad hoc logic.
Like we did here  https://github.com/getsentry/snuba/blob/master/snuba/query/parser/expressions.py#L18-L43
It would save you all the manual regex matching, and all the custom logic to decide decide when to look for a modifier, which is very hard to verify.

lynnagara replied:
I will give it a shot to see how it compares

fpacifici commented on snuba/clickhouse/columns.py:
```diff
@@ -90,6 +85,11 @@ def get_nested_modifiers(
+    def __init__(self, inner_type: ColumnType) -> None:
+        self.inner_type = inner_type
+
+
```
This seems wrong now, We have a ColumnTypeWithModifier, but get_all_modifiers is outside of it, seems this is a leaky abstraction. I assume you did this to avoid depending on this file form errors_replacer, I guess for a circular dependency. Could you please expand on what the reason was ?

lynnagara replied:
The reason is doing `is_instance(x, ModifiedColumnType and LowCardinality in x.get_all_modifiers()` everywhere to figure out if a given column is low cardinality seems tedious. Given a column type it seems like it should be easier to find the list of modifiers. Would changing the implementation help so ColumnType doesn't reference ColumnTypeWithModifier help? I just did it as a straight copy/paste to get this working quickly for now.

fpacifici replied:
Can we then split this implementation? Let the general get_all_modifiers return [] and have a specific implementation for ColumnTypeWithModifier that provides the recursive implementation? 
That way ColumnTypeWithModufier actually has some responsibility and remove the isinstance calls.

lynnagara replied:
Yep, moved this change into https://github.com/getsentry/snuba/pull/803

fpacifici reviewed (changes requested):
I'd really invest some time in writing a proper grammar to parse the clickhouse type expressions. It would generate a much more maintainable and robust logic.

lynnagara committed:
get_all_modifierrs

lynnagara committed:
reimplement with grammar

lynnagara committed:
Merge remote-tracking branch 'origin/master' into fix-migrate-errors

fpacifici reviewed (commented):
The grammar is a good step ahead, please add unit tests for that logic, it is easy to test since it is stateless and, since there are a lot of possible cases you can easily create many cases in this form:
https://github.com/getsentry/snuba/blob/master/tests/query/parser/test_expressions.py

Otherwise validating the grammar and avoiding regressions will be a pain due to the amount of cases to test.

lynnagara committed:
move things around

lynnagara committed:
fix import error

lynnagara committed:
add test

lynnagara committed:
Merge remote-tracking branch 'origin/master' into fix-migrate-errors

lynnagara committed:
match single comma space

lynnagara committed:
improve grammar

lynnagara commented on snuba/migrate.py:
```diff
@@ -1,24 +1,163 @@
+    lowcardinality   = "LowCardinality" inner_type
+    nullable         = "Nullable" inner_type
+    array            = "Array" inner_type
+    inner_type       = ~r"\([a-zA-Z0-9()]+\)"
```
Done, looks a lot clearer this way

lynnagara commented on snuba/migrate.py:
```diff
@@ -1,24 +1,163 @@
+    fixedstring_size = ~r"\d+"
+    enum             = "Enum" enum_size open_paren enum_pairs close_paren
+    enum_size        = "8" / "16"
+    enum_pairs       = (enum_pair comma* space*)*
```
Fixed. It doesn't cover some edge cases still (like absence of any ", " between pairs) but we shouldn't get invalid data in here and I haven't seen a concise way to accomplish this

lynnagara committed:
Merge remote-tracking branch 'origin/master' into fix-migrate-errors

lynnagara commented:
@fpacifici Made your requested changes, any thoughts about this now?

fpacifici commented:
I think this is still missing a good test coverage for the parser. I thought you were working on that. 
The issue I have in just  testing the difference between the local schema and the dataset schema is that it only tests the outcome of the migrations we happen to have now.
As a result we don't get a solid confidence that the grammar is right and it is properly processed in corner cases, let alone testing regressions on that grammar. Thus having a specific unit test for the grammar that test an extensive amount of cases (including corner cases) as we did for the query parsing. This gives you a good sense that the grammar is not regressing without depending on the specific dataset migrations we happen to have.

fpacifici reviewed (changes requested):
back to you for test coverage

lynnagara committed:
fix: Default expressions on column should always be strings

Default expressions are always strings, Clickhouse will cast values to
the appropriate type for that column.

lynnagara committed:
add test

lynnagara committed:
Merge branch 'master' into fix-migrate-errors

fpacifici commented on snuba/datasets/schemas/__init__.py:
```diff
@@ -107,6 +109,8 @@ def get_column_differences(self, expected_columns: Mapping[str, str]) -> List[st
             expected_type = self.get_columns()[column_name].type.for_schema()
+
+            column_type = column.for_schema()
             if column_type != expected_type:
```
Any reason to compare the string representation in stead of the Type object now that we deal with ColumnType classes around ? I think we should turn this into a ColumnType comparison if it is possible.

lynnagara replied:
Not really, done

fpacifici commented on snuba/migrations/migrate.py:
```diff
@@ -45,7 +32,6 @@ def run(conn, dataset):
     schemas = []
     if dataset.get_table_writer():
         schemas.append(dataset.get_table_writer().get_schema())
-    schemas.append(dataset.get_dataset_schemas().get_read_schema())
```
Why this removal ? If the datasets has multiple schemas out of which several read ones and one writer (like matviews that have 2 read schemas and a write schema) the read schemas would not be migrated.

lynnagara replied:
Yeah good point, I think this was an oversight.

fpacifici commented on snuba/migrations/parse_schema.py:
```diff
@@ -0,0 +1,152 @@
+    uint_size        = "8" / "16" / "32" / "64"
+    float            = "Float" float_size
+    float_size       = "32" / "64"
+    fixedstring      = "FixedString" open_paren fixedstring_size close_paren
```
You should always leave an optional sequence of spaces after ( and before ) and  before and after commas. Clickhosue is probably not going to format its expressions by adding spaces but they would be valid expressions

lynnagara replied:
Makes sense, done

fpacifici commented on snuba/migrations/parse_schema.py:
```diff
@@ -0,0 +1,152 @@
+    fixedstring_size = ~r"\d+"
+    enum             = "Enum" enum_size open_paren enum_pairs close_paren
+    enum_size        = "8" / "16"
+    enum_pairs       = (enum_pair (comma space)?)*
```
There can be potentially multiple spaces after the comma

fpacifici commented on tests/migrations/test_parse_schema.py:
```diff
@@ -0,0 +1,62 @@
+from snuba.migrations.parse_schema import _get_column
+
+
+test_data = [
```
nice. thanks.

fpacifici reviewed (changes requested):
Almost there. I think there is an issue in the function that selects the schemas to migrate that now discards read schemas.

lynnagara committed:
replace migrate read schemas

lynnagara committed:
aggregate function basic support

fpacifici commented:
Please fix the test and see the other comments from my previous reviews before merging

lynnagara committed:
support expressions with optional spaces

lynnagara committed:
Merge branch 'master' into fix-migrate-errors

lynnagara committed:
compare ColumnType objects
